[aten] Use emplace instead of insert.

ttsugriy opened the pull request:
Summary:
this avoids creation of a pair in case inserted key
is already in dict

Test Plan: CI

Reviewed By: swolchok

Differential Revision: D52641840

Skylion007 commented on aten/src/ATen/core/Dict_inl.h:
```diff
@@ -120,9 +120,9 @@ template<class Key_, class Value_>
-  auto inserted = impl_->dict.insert(std::pair<IValue, IValue>{
-    Key(std::forward<Key_>(key)),
-    Value(std::forward<Value_>(value))});
+  auto inserted = impl_->dict.emplace(
```
```suggestion
  auto inserted = impl_->dict.try_emplace(
```
Pretty sure this is what you intended. This will avoid creating the the std::pair entirely if it's already in the dict.

ttsugriy replied:
nope, that's most definitely not what I intended ;) There are at least 2 reasons for this:
1) `try_emplace` exists only in broken implementations of dict like the ones that are in std. Such function makes no sense in non-broken implementations that aten is using https://github.com/pytorch/pytorch/blob/main/c10/util/order_preserving_flat_hash_map.h#L537-L553
2) the implementation from aten already skips pair creation if key is present

ttsugriy committed:
[aten] Use emplace instead of insert. (#117283)

Summary:

this avoids creation of a pair in case inserted key
is already in dict

Test Plan: CI

Reviewed By: swolchok

Differential Revision: D52641840
